hudi-flink/src/main/java/org/apache/hudi/operator/StreamWriteFunction.java

```diff
@@ -0,0 +1,344 @@
+ * When the existing data buffer write successfully, the process thread unblock and start buffering again for the next round checkpoint.
+ * Because any checkpoint failures would trigger the write rollback, it implements the exactly-once semantics.
+ *
+ * <p><h2>Fault Tolerance</h2>
+ *
+ * <p>The operator coordinator checks the validity for the last instant when it starts a new one. The operator rolls back
+ * the written data and throws when any error occurs. This means any checkpoint or task failure would trigger a failover.
+ * The operator coordinator would try several times when committing the writestatus.
+ *
+ * <p>Note: The function task requires the input stream be partitioned by the partition fields to avoid different write tasks
+ * write to the same file group that conflict. The general case for partition path is a datetime field,
+ * so the sink task is very possible to have IO bottleneck, the more flexible solution is to shuffle the
+ * data by the file group IDs.
+ *
+ * @param <I> Type of the input record
+ * @see StreamWriteOperatorCoordinator
+ */
+public class StreamWriteFunction<K, I, O> extends KeyedProcessFunction<K, I, O> implements CheckpointedFunction {
+
+  private static final long serialVersionUID = 1L;
+
+  private static final Logger LOG = LoggerFactory.getLogger(StreamWriteFunction.class);
+
+  /**
+   * Write buffer for a checkpoint.
+   */
+  private transient List<HoodieRecord> buffer;
+
+  /**
+   * The buffer lock to control data buffering/flushing.
+   */
+  private transient ReentrantLock bufferLock;
+
+  /**
+   * The condition to decide whether to add new records into the buffer.
+   */
+  private transient Condition addToBufferCondition;
+
+  /**
+   * Flag saying whether there is an on-going checkpoint.
+   */
+  private volatile boolean onCheckpointing = false;
+
+  /**
+   * Config options.
+   */
+  private final Configuration config;
+
+  /**
+   * Id of current subtask.
+   */
+  private int taskID;
+
+  /**
+   * Write Client.
+   */
+  private transient HoodieFlinkWriteClient writeClient;
+
+  private transient BiFunction<List<HoodieRecord>, String, List<WriteStatus>> writeFunction;
+
+  /**
+   * HoodieKey generator.
+   */
+  private transient KeyGenerator keyGenerator;
+
+  /**
+   * Row type of the input.
+   */
+  private final RowType rowType;
+
+  /**
+   * Avro schema of the input.
+   */
+  private final Schema avroSchema;
+
+  private transient RowDataToAvroConverters.RowDataToAvroConverter converter;
+
+  /**
+   * The REQUESTED instant we write the data.
+   */
+  private volatile String currentInstant;
+
+  /**
+   * Gateway to send operator events to the operator coordinator.
+   */
+  private transient OperatorEventGateway eventGateway;
+
+  /**
+   * Constructs a StreamingSinkFunction.
+   *
+   * @param rowType The input row type
+   * @param config  The config options
+   */
+  public StreamWriteFunction(RowType rowType, Configuration config) {
```

Comment:
Actually, do not know we really need to use `RowType`, only for convenience to integrate with `table/sql`?

Reply:
Use row type has 2 benefits:
1. Use the all kinds of `DeserializationSchema` provided for `RowData`, such as `JsonRowDataDeserializationSchema`, there is no need to write converter code for every data format, `RowData` is a good data structure to bridge external raw format with `HoodieRecord`
2. `RowData` is internal data structure of `SQL` API, so that the codes for datastream can be reused.

Reply:
Yes, I know this, but still brings external dependency for the DataStream users.  There are two concerns:

1) This brings an additional burden for sharing the same abstraction with the current implementation;
2) RowType makes the number of compatible Flink versions less.

Reply:
> This brings an additional burden for sharing the same abstraction
The main abstraction that we can share is the partitioning and micro-batching, not the source record parsing, so this should not be a concern. On the contrary, the `RowData` is very efficient in processing and it supports many data formats, e.g. `JSON`/`CSV`/`Avro` which are all complex and have many config options.

> RowType makes the number of compatible Flink versions less
The old pipeline does not use the `RowType`, so it should not have compatibility problem.

BTW, the Hoodie `Spark` data source also use `Row` as the basic data structure, which is easy for `JSON`/`CSV`/`Avro` source data conversion and effective data handling in the pipeline.

Reply:
> The main abstraction that we can share is the partitioning and micro-batching, not the source record parsing, so this should not be a concern.

I mean if two write clients use the same basic data structure, there would be more logic that could be put into the abstraction.